lib/logstash/inputs/beats.rb

```diff
@@ -189,6 +206,22 @@ def register
       @logger.warn("configured ssl_key => #{@ssl_key.inspect} will not be used") if @ssl_key
     end
 
+    active_enrichments = resolve_enriches
+
+    @include_source_metadata = active_enrichments.include?('source_metadata')
+    @include_codec_tag = original_params.include?('include_codec_tag') ? params['include_codec_tag'] : active_enrichments.include?('codec_metadata')
+    @ssl_peer_metadata = original_params.include?('ssl_peer_metadata') ? params['ssl_peer_metadata'] : active_enrichments.include?('ssl_peer_metadata')
+
+    # intentionally ask users to provide codec when they want to use the codec metadata
+    # second layer enrich is also a controller, provide enrich => ['codec_metadata' or/with 'source_metadata'] with codec if you override event original
+    unless active_enrichments.include?('codec_metadata')
+      if original_params.include?('codec')
+        @logger.warn('`enrich` configuration does not include `codec_metadata`, but the directive is not propagated to the explicitly-defined `codec`')
```

Comment:
As a non native English speaker it's hard for me gauge the consequences of this warning. In fact, getting this warning after setting "codec_metadata" in enrich w/ an explicit codec suggests we won't be skipping the  `beats_input_codec_plain_applied`/`beats_input_raw_event` tags, but we do (see `enrich => none codec => plain { charset => 'ISO8859-1' } }` example on my big comment). However we don't warn about not disabling ECS compatibility anymore.

I believe we should have a clearer relationship between ECS compatibility and the codec_metadata enrich option.
I see a few alternatives:
1. creating a separate enrich option just for the ECS toggling
2. expanding codec_metadata to be....`codec_metadata_and_ecs`?

Reply:
Warning message: `the directive is not propagated` statement would be a reason for misunderstanding (or not understanding). Paraphrasing with some options where we can choose one of them or give your options please:
- "`enrich` configuration does not include `codec_metadata`. Explicitly defined `codec` will apply ECS compatibility by default or align on `codec` `ecs_compatibility` setting."
- "`codec` is specified but `enrich` configuration does not include `codec_metadata`. ECS compatibility is enabled by default or align on `codec` `ecs_compatibility` configuration."
- "`codec_metadata` option is not include in `enrich` configuration. This will lead the `codec` to control ECS compatibility (by default enabled) based on its `ecs_compatibility` configuration."

1. As my understanding, we have `ecs_compatibility` which gives full control codec's ECS behavior and we have now `codec_metadata` to set the codec behavior. I need more information about the enrich option you are mentioning and how it should work, its relations with currently available settings (such as `ecs_compatibility`), etc...
2. `codec_metadata` -> `codec_metadata_and_ecs`
 (I understood expanding codec_metadata to be `codec_metadata_and_ecs` as a naming change - plz correct me if its meaning is different)
It is true the currently we are mostly referring to `[event][original]` ECS behavior case but `codec_metadata`, a broader naming would give us extra room to control metadata cases in the future.

Reply:
I think there is a challenge in explaining the interaction between `enrich => codec_metadata` and `codec`:

If no `codec` it specified and:
a) `enrich => codec_metadata`: plugin adds codec tags and does not influence `ecs_compatibility`
b) `enrich => none`: plugin does not add codec tags and does not influence `ecs_compatibility`

If `codec` is not specified and:
a) `enrich => codec_metadata`: plugin adds codec tags does not influence `ecs_compatibility`
a) `enrich => none`: plugin does not add codec tags **and disables `ecs_compatibility`**

Two potential surprises when `ecs_compatibility` isn't influenced:
1) it's possible for `beats { port => 3333 enrich => codec_metadata }` to not add [event][original], if `pipeline.ecs_compatibility: disabled` is configured in the logstash.yml.
2) Contrary, it's possible for `enrich => none` to still add [event][original] if a codec is set.

I don't think there's a single great option here, unfortunately. The disabling of ECS compatibility through enrich is a very useful convenience but it doesn't fit 100% into the logic of the settings we provide. The next best thing really comes down to providing enough documentation and logging of setting behaviour.
Maybe in the future we can just agree that `ecs_compatibility` does not make sense for this plugin.

Until then we can go with one of the messages you described. Maybe a little tweak:

> "codec is specified but enrich configuration does not include codec_metadata. ECS compatibility is enabled by default or align on codec ecs_compatibility configuration."

We can also add a debug logging entry for the other branch of the conditional:

> "disabling 'ecs_compatibilty' for the default codec since 'enrich' configuration does not include 'codec_metadata' and no explicit codec is set"

Reply:
I think you mean by `If _no_ codec it specified and:` is `If codec is specified and:`, right? 

I have slightly improved the doc and the logger messages with [this commit](https://github.com/logstash-plugins/logstash-input-beats/pull/464/commits/548e99e2cd58ee80beebd0d5bca52d6a27ef0fae) as your per request and I TOTALLY AGREE with the opinion of **_Maybe in the future we can just agree that `ecs_compatibility` does not make sense for this plugin_** if we think from Elastic stack (+beats, agent) point of view. That was the reason why I initially [pushed _no-transformation_ idea](https://github.com/logstash-plugins/logstash-input-beats/pull/464#discussion_r1064348212). However, this would be a breaking change.

Reply:
> I think you mean by If _no_ codec it specified and: is If codec is specified and:, right?

correct. fixed.

> That was the reason why I initially https://github.com/logstash-plugins/logstash-input-beats/pull/464#discussion_r1064348212. However, this would be a breaking change.

Indeed, small steps. with `enrich` we have a door towards influencing event manipulation. And we can perhaps decide to deprecate ecs_compatibilty alongside other settings like the ssl and codec enrichments.